huaweicloud/services/iotda/resource_huaweicloud_iotda_dataforwarding_rule.go

```diff
@@ -0,0 +1,732 @@
+					ActionId: id.(string),
+					Body: &model.UpdateActionReq{
+						Channel:       &channel,
+						ChannelDetail: channelDetail,
+					},
+				})
+				if err != nil {
+					return diag.Errorf("error updating targets of IoTDA data forwarding rule: %s", err)
+				}
+			} else {
+				_, err = client.CreateRuleAction(&model.CreateRuleActionRequest{
+					Body: &model.AddActionReq{
+						RuleId:        d.Id(),
+						Channel:       channel,
+						ChannelDetail: channelDetail,
+					},
+				})
+				if err != nil {
+					return diag.Errorf("error updating targets of IoTDA data forwarding rule: %s", err)
+				}
+			}
+
+		}
+	}
+
+	// This update must be the last
+	if d.HasChanges("name", "description", "select", "where", "enabled") {
+		_, err = client.UpdateRoutingRule(&model.UpdateRoutingRuleRequest{
+			RuleId: d.Id(),
+			Body: &model.UpdateRuleReq{
+				RuleName:    utils.String(d.Get("name").(string)),
+				Description: utils.StringIgnoreEmpty(d.Get("description").(string)),
+				Select:      utils.StringIgnoreEmpty(d.Get("select").(string)),
+				Where:       utils.StringIgnoreEmpty(d.Get("where").(string)),
+				Active:      utils.Bool(d.Get("enabled").(bool)),
+			},
+		})
+
+		if err != nil {
+			return diag.Errorf("error updating IoTDA data forwarding rule: %s", err)
+		}
+	}
+
+	return ResourceDataForwardingRuleRead(ctx, d, meta)
+}
+
+func ResourceDataForwardingRuleDelete(_ context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	c := meta.(*config.Config)
+	region := c.GetRegion(d)
+	client, err := c.HcIoTdaV5Client(region)
+	if err != nil {
+		return diag.Errorf("error creating IoTDA v5 client: %s", err)
+	}
+
+	//delete targets
+	targets := d.Get("targets").(*schema.Set)
+	for _, v := range targets.List() {
+		ruleAction := v.(map[string]interface{})
+		_, err = client.DeleteRuleAction(&model.DeleteRuleActionRequest{ActionId: ruleAction["id"].(string)})
+		if err != nil {
+			return diag.Errorf("error deleting targets of IoTDA data forwarding rule: %s", err)
+		}
+	}
+
+	// delete data forwarding
+	deleteOpts := &model.DeleteRoutingRuleRequest{RuleId: d.Id()}
+	_, err = client.DeleteRoutingRule(deleteOpts)
+	if err != nil {
+		return diag.Errorf("error deleting IoTDA data forwarding rule: %s", err)
+	}
+
+	return nil
+}
+
+func buildDataForwardingRuleCreateParams(d *schema.ResourceData) *model.CreateRoutingRuleRequest {
+	triggers := strings.SplitN(d.Get("trigger").(string), ":", 2)
+	req := model.CreateRoutingRuleRequest{
+		Body: &model.AddRuleReq{
+			RuleName:    utils.String(d.Get("name").(string)),
+			Description: utils.StringIgnoreEmpty(d.Get("description").(string)),
+			Select:      utils.StringIgnoreEmpty(d.Get("select").(string)),
+			Where:       utils.StringIgnoreEmpty(d.Get("where").(string)),
+			AppId:       utils.StringIgnoreEmpty(d.Get("space_id").(string)),
+			Subject: &model.RoutingRuleSubject{
+				Resource: triggers[0],
+				Event:    triggers[1],
```

Comment:
suggest to check the array index to avoid panic.

Reply:
checked in schema